### PR TITLE
Contribute glsp-tutorial and configure gitpod

### LINKED ISSUE #2: Switch to plain minmal example
Define a tag and checkout this version

### LINKED ISSUE #14: Contribute GLSP tutorial
To repo either inside GLSP or eclipsesource.

### PR DESCRIPTION
**Pull request by max-elia**
* Add all the files and gitpod configuration for loading the tutorial files together with the theia-tutorial-extension
* Change Dockerfile to choose lower node version since newer node versions are not compatible with the glsp-tutorial
* Closes eclipsesource/theia-tutorial-extension#14
* Switch to glsp-example repo without assistance
* Host images on github
* Update Year
* Change test folder structure
* Change prebuild task
* Update Readme
* Update .tut.json to adjust to changes
Closes #2

Signed-off-by: Max Elia Schweigkofler <max_elia@hotmail.de>

### CODE REVIEW

File: .tutorial/solution/di.config.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/********************************************************************************
+ * Copyright (c) 2020 EclipseSource and others.
```
Comment (eneufeld) [inline .tutorial/solution/di.config.ts]: The copyright should probably be 2021 ?

File: .tutorial/glsp.tut.json [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+			"title": "Introduction and set-up",
+			"description": "Getting started learning the basics of the Eclipse Graphical Language Server Platform (GLSP) and setting up the workspace for the tutorial.",
+			"typeId": "#exercise",
+			"content": [
+				{
+					"html": "<h2>Eclipse Graphical Language Server Platform (GLSP)</h2><br/><p>The Graphical Language Server Platform provides extensible components for the development of diagram editors including edit functionality in (distributed) web-applications via a client-server protocol.</p>"
+				},
+				{
+					"image": {
+						"src": "https://cdn.discordapp.com/attachments/776069960371273769/803300788737933403/glsp1.gif",
```
Comment (eneufeld) [inline .tutorial/glsp.tut.json]: is it the best way to retrieve images from discord?
Shouldn't we add an image directory in this repo and take the images from there?

**Review by eneufeld — COMMENTED** [review]
I assume this is based on this Example: https://github.com/eclipse-glsp/glsp-examples/tree/master/minimal ?
We probably should link this in the readme or at least mention it.
Can we copy the necessary files for the solutions from the external repo or do we need them to be in the `.tutorial` folder?

**Review by eneufeld — COMMENTED** [review]
Why is `.tutorial/test/java/org/eclipse/glsp/example/minimal/handler/TestMinimalCreateNodeOperationHandler.java` much deeper nested than all other files?

File: .tutorial/solution/di.config.ts (hunk 1 above)
Comment (max-elia) [reply .tutorial/solution/di.config.ts]: Changed it

File: .tutorial/glsp.tut.json (hunk 2 above)
Comment (max-elia) [reply .tutorial/glsp.tut.json]: Created image directory

Comment (max-elia) [reply .tutorial/glsp.tut.json]: Created image directory

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import {injectable, inject} from "inversify";
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: I think it makes sense to apply the prettier and eslint rules to this repo, too

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import {WorkspaceService} from '@theia/workspace/lib/browser';
+
+
+export const AssistanceCommand = {
+    id: 'Assistance.command',
+    label: "assistance"
+};
+
+@injectable()
+export class markingElements implements FrontendApplicationContribution {
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: Classes usually start with an Uppercase letter, I suggest to keep this

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: here are some unnecessary empty lines, maybe remove them?

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+    idList: Array<string> = [];
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: do all those members need to be public?

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    currentHint: number = -1;
+    observer: MutationObserver | null;
+    leftPostion: number;
+    topPosition: number;
+
+    constructor(
+
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: empty line

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        @inject(FileService) private readonly fileService: FileService,
+        @inject(WorkspaceService) private readonly workspaceService: WorkspaceService
+    ) { }
+
+    findNewCurrent = () => {
+        for (let i = 0; i < this.idList.length; i++) {
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: you could do:
```
this.idList.forEach((tmp,i) => {...})
```
No need to change the for loop right now, just keep this in mind

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                    .filter(function () {return $(this).children().length === 0 && $(this).text() === tmp;})
+                    .parent().length > 0)) {
+                this.currentHint = i;
+                this.currentLeftPosition();
+                this.currentTopPosition();
+            }
+        }
+    }
+
+    asId = (id: string) => {
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: here are some methods which are actually util functions and would work standalone.
I suggest to create a util file and move all those functions there. Just declare them like this:
```
export const asId = (id: string) => {...};
```

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        return "#" + $.escapeSelector(id);
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: do we really need jquery?
I'm fine using it if the code is shorter and cleaner with it, but usually it doesn't make much difference

**Review by eneufeld — COMMENTED** [review]
This looks good. 
We should apply formatting and linting rules to the repo, though.

File: .tutorial/glsp.tut.json [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
...implemented using HTML5, Typescript, Sprotty and SVG. The generic GLSP diagram client can be connected to any domain-specific GLSP server via the defined protocol to support a specific diagram type. Furthermore, the diagram client can be customized and integrated in a website, a VS Code CLient, a Theia Client or run as a standalone Client.</li></ul><li>Graphical Language Server Protocol</li><ul><li>Following the successful pattern provided by the language server protocol (LSP) for textual languages, GLSP provides a defined protocol for the communication between the GLSP client and the server. In this modular architecture, server and client are well encapsulated and can be developed in an independant way.</li></ul><li>Graphical Language Server</li><ul><li>Enabled by the flexible protocol-based architecture, a domain-specific GLSP server can be implemented in any preferred technology. Therefore, you can easily integrate a GLSP server into an existing environment and reuse the domain-specific business and diagram logic. GLSP provides a frame for the implementation of GLSP servers using Java, as well as integrations with EMF and other existing modeling technologies.</li></ul></ul></p>"
```
Comment (eneufeld) [inline .tutorial/glsp.tut.json]:
```suggestion
					"html": "<p>We will have a closer look at the components:<ul><li>Graphical Model Editor Client</li><ul><li>The client is implemented using HTML5, Typescript, Sprotty and SVG. The generic GLSP diagram client can be connected to any domain-specific GLSP server via the defined protocol to support a specific diagram type. Furthermore, the diagram client can be customized and integrated in a website, a VS Code CLient, a Theia Client or run as a standalone Client.</li></ul><li>Graphical Language Server Protocol</li><ul><li>Following the successful pattern provided by the language server protocol (LSP) for textual languages, GLSP provides a defined protocol for the communication between the GLSP client and the server. In this modular architecture, server and client are well encapsulated and can be developed in an independant way.</li></ul><li>Graphical Language Server</li><ul><li>Enabled by the flexible protocol-based architecture, a domain-specific GLSP server can be implemented in any preferred technology. Therefore, you can easily integrate a GLSP server into an existing environment and reuse the domain-specific business and diagram logic. GLSP provides a frame for the implementation of GLSP servers using Java, as well as integrations with EMF and other existing modeling technologies.</li></ul></ul></p>"
```

**Review by eneufeld — APPROVED** [review]
Looks good, maybe we should squash the commits.
I also added a suggestion for a language fix.

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+          .parent().length > 0
+      ) {
+        this.currentHint = i;
+        this.currentLeftPosition();
+        this.currentTopPosition();
+      }
+    }
+  };
+
+  asId = (id: string) => {
```
Comment (eneufeld) [inline .tutorial/assistance/assistance/src/browser/assistance-contribution.ts:46]: I still suggest to move all methods which can `static` to an util class.
also double check the necessity of jquery.
I'm fine if you create a ticket for this to tackle both issues in the future

**Review by eneufeld — APPROVED** [review]
Thank you. I still have a potential improvement suggestion, which can be tackled in the future though.

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts (hunk 10 above)
Comment (max-elia) [reply .tutorial/assistance/assistance/src/browser/assistance-contribution.ts]: Created issue #5

File: .tutorial/assistance/assistance/src/browser/assistance-contribution.ts (hunk 12 above)
Comment (max-elia) [reply .tutorial/assistance/assistance/src/browser/assistance-contribution.ts:46]: Included in issue #5

**Pull request merged**